[OCPNODE-852] workerlatency profiles - updating the kubelet configuration based on the nodes.config.openshift.io resource

sairameshv opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->

**- What I did**
Added support to update the kubelet configuration based on the node.config.openshift.io spec changes

**-How to verify it**
Create/Update the nodes.config.openshift.io CR with the relevant/valid workerlatency profile and expect the corresponding changes in the Kubelet Configuration that are rolled out on all the worker nodes by MCO leveraging the KubeletConfig CR.

**- Description for the changelog**
Based on the [enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/worker-latency-profile/worker-latency-profile.md), users can create/modify the node.config.openshift.io resource which results in the corresponding  kubelet configuration changes on all the worker nodes.
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

sairameshv commented:
/hold
Based on the discussions happened regarding https://github.com/openshift/machine-config-operator/pull/2959, these changes are made as part of an alternative approach.
cc: @harche @rphillips @yuqi-zhang

sairameshv commented:
/retest-required

rphillips commented on pkg/apis/machineconfiguration.openshift.io/v1/types.go:
```diff
@@ -13,6 +13,9 @@ import (
 const MachineConfigRoleLabelKey = "machineconfiguration.openshift.io/role"
 
+// KubeletConfigRoleLabel is the label that must be present in the KubeletConfig CR
+const KubeletConfigRoleLabel = "pools.operator.machineconfiguration.openshift.io/"
```
perhaps call this KubeletConfigRoleLabelPrefix since it is only the prefix.

rphillips commented on pkg/controller/kubelet-config/helpers.go:
```diff
@@ -131,6 +133,40 @@ func createNewDefaultFeatureGate() *osev1.FeatureGate {
+		// (TODO): The durations can be replaced with the defined constants in the openshift/api repository once the respective changes are merged.
+		switch node.Spec.WorkerLatencyProfile {
+		case osev1.MediumUpdateAverageReaction:
+			originalKubeletConfig.NodeStatusUpdateFrequency = metav1.Duration{Duration: 20 * time.Second}
```
these should be constants.

rphillips commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,276 @@
+		glog.V(2).Infof("%v", err)
+		err := fmt.Errorf("could not fetch Node: %v", err)
+		return err
+	} else if err == nil {
```
can be just else {} ... instead of the if..

rphillips commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,276 @@
+	for _, pool := range mcpPools {
+		role := pool.Name
+		// the configuration change will be applied only on the worker nodes.
+		if role == "master" {
```
there should be a constant for this in the MCO code.

rphillips commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,276 @@
+		if err != nil {
+			return err
+		}
+		if byteData == nil {
```
this if clause is impossible... it can be removed.

rphillips commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,276 @@
+// isStatusReady returns a boolean based on the condition statuses present in the Node CR status.
+// If the status condition type is Progressing or Degraded, this function returns false
+func isStatusReady(node *osev1.Node) bool {
+	if node != nil {
```
reverse the logic here to avoid the nested block

rphillips commented:
```
test/e2e-shared-tests/helpers.go:167:19: ambiguous selector cs.Nodes (typecheck)
		node, err := cs.Nodes().Get(context.TODO(), target.Name, metav1.GetOptions{})
		                ^
```

QiWang19 commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,274 @@
+			return err
+		}
+
+		kc, err := ctrl.client.MachineconfigurationV1().KubeletConfigs().Get(context.TODO(), managedKey, metav1.GetOptions{})
```
use argument `key` instead of `managedKey` since it's getting kubeletconfigs.

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,273 @@
+	for _, pool := range mcpPools {
+		role := pool.Name
+		// the configuration change will be applied only on the worker nodes.
+		if role == string(mcfgv1.MachineConfigPoolMaster) {
```
why string() here?

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,273 @@
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		isNotFound := errors.IsNotFound(err)
```
move this up and use it in if above

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,273 @@
+// updateNodestatus updates the status of the node object based on the machineconfigpool status
+func (ctrl *Controller) updateNodestatus(kc *mcfgv1.KubeletConfig, node *osev1.Node) error {
+	if kc == nil || node == nil {
+		return fmt.Errorf("unable to update the node status, incomplete data, kubelet config: %v, node: %v", kc, node)
```
what does node = nil mean in this case? why call updateNodestatus if node is nil in the first place?

sairameshv replied:
Just to avoid any unwanted panics in case of accessing a nil object, this check is added.

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,273 @@
+}
+
+// updateNodestatus updates the status of the node object based on the machineconfigpool status
+func (ctrl *Controller) updateNodestatus(kc *mcfgv1.KubeletConfig, node *osev1.Node) error {
```
```suggestion
func (ctrl *Controller) updateNodeStatus(kc *mcfgv1.KubeletConfig, node *osev1.Node) error {
```

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/helpers.go:
```diff
@@ -131,6 +132,40 @@ func createNewDefaultFeatureGate() *osev1.FeatureGate {
+}
+
+// updateOriginalKubeConfigwithNodeConfig updates the original Kubelet Configuration based on the Nodespecific configuration
+func updateOriginalKubeConfigwithNodeConfig(node *osev1.Node, originalKubeletConfig *kubeletconfigv1beta1.KubeletConfiguration) error {
```
is osev1.Node a node config? If so can we name the variables indicating that? It's a little confusing reading through..

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_bootstrap.go:
```diff
@@ -43,6 +43,11 @@ func RunKubeletBootstrap(templateDir string, kubeletConfigs []*mcfgv1.KubeletCon
 				return nil, err
 			}
+			// updating the original kubelet configuration with the Node specific configuration.
+			if node != nil {
```
maybe nodeConfig here?

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,273 @@
+
+func (ctrl *Controller) addNode(obj interface{}) {
+	node := obj.(*osev1.Node)
+	glog.V(4).Infof("Add Node event %s", node.Name)
```
Same "Add Node event : node name" isn't really clear for debugging

sairameshv replied:
Addressed

kikisdeliveryservice replied:
nice change :+1:

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,273 @@
+	ctrl.enqueueNode(node)
+}
+
+func (ctrl *Controller) deleteNode(obj interface{}) {
```
I'm a little confused overall, but what's happening here? Are you deleting a Node or deleting a node config??

sairameshv replied:
Addressed

kikisdeliveryservice reviewed (commented):
A couple of questions/comments, mostly, this Node reference is also really confusing to me and when you're talking about a Node config vs a node isn't quite obvious especially since in the MCO we do deal with nodes.. I can't quite figure out overall what's happening throught bc of it. Is there room to clarify within the code?

sairameshv commented:
> A couple of questions/comments, mostly, this Node reference is also really confusing to me and when you're talking about a Node config vs a node isn't quite obvious especially since in the MCO we do deal with nodes.. I can't quite figure out overall what's happening throught bc of it. Is there room to clarify within the code?

Added the comments in the code with an appropriate description [here](https://github.com/sairameshv/machine-config-operator/blob/node_object/pkg/controller/kubelet-config/kubelet_config_nodes.go#L60-L62)

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,274 @@
+			glog.Infof("Skipping the Update Node event, name: %s, transition not allowed from old WorkerLatencyProfile: %s to new WorkerLatencyProfile: %s", newNode.Name, oldNode.Spec.WorkerLatencyProfile, newNode.Spec.WorkerLatencyProfile)
+			return
+		}
+		glog.V(4).Infof("Update Node event, name: %s", newNode.Name)
```
Addressed

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,274 @@
+// isStatusReady returns a boolean based on the condition statuses present in the Node CR status.
+// If the status condition type is Progressing or Degraded, this function returns false
+func isStatusReady(node *osev1.Node) bool {
+	if node == nil {
```
Updated the method name along with appropriate description. On a high-level, this method helps to decide whether to act or not, upon receiving any update on the node config object.

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,274 @@
+	return true
+}
+
+// updateNodestatus updates the status of the node object based on the machineconfigpool status
```
Addressed

sairameshv commented on pkg/apis/machineconfiguration.openshift.io/v1/types.go:
```diff
@@ -13,6 +13,9 @@ import (
 const MachineConfigRoleLabelKey = "machineconfiguration.openshift.io/role"
 
+// KubeletConfigRoleLabelPrefix is the label that must be present in the KubeletConfig CR
+const KubeletConfigRoleLabelPrefix = "pools.operator.machineconfiguration.openshift.io/"
```
Based on the following example(https://github.com/openshift/machine-config-operator/blob/master/docs/KubeletConfigDesign.md#example---setting-the-kubelet-log-level), I understood that the `pools.operator.machineconfiguration.openshift.io/{worker|master}: ""` label must be present in the KubeletConfig CR inorder to get picked up by the respective pool `worker|master`. Hence I have defined a constant here for the prefix so that it can be used accordingly.

kikisdeliveryservice replied:
gah sorry, this is fine, it needs to be used by controller manager/apiserver/etc...

sairameshv commented:
@kikisdeliveryservice , I have addressed the review comments, could you PTAL?

rphillips commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,336 @@
+}
+
+func (ctrl *Controller) addNodeConfig(obj interface{}) {
+	nodeConfig := obj.(*osev1.Node)
```
We should be defensive here and use the `nodeConfig, ok` return to make sure it is a osev1.Node object.

sairameshv replied:
Sure, addressed

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_bootstrap.go:
```diff
@@ -43,6 +43,7 @@ func RunKubeletBootstrap(templateDir string, kubeletConfigs []*mcfgv1.KubeletCon
 			if err != nil {
 				return nil, err
 			}
+
```
Is it my view or is the only change in this file a blank line now (there used to be code here)? If just a blank line, let's remove that. :)

sairameshv replied:
Addressed

kikisdeliveryservice commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,336 @@
+	return ctrl.updataMCOStatus(nodeConfig, nodeConfigCondition)
+}
+
+// updataMCOStatus updates the MCO specific status condition present in the node config CR
```
```suggestion
// updateMCOStatus updates the MCO specific status condition present in the node config CR
```

sairameshv replied:
Addressed

yuqi-zhang commented on pkg/controller/common/helpers.go:
```diff
@@ -668,6 +668,35 @@ func MachineConfigFromRawIgnConfig(role, name string, rawIgnCfg []byte) (*mcfgv1
+// KubeletConfigFromRawIgnConfig creates a KubeletConfig with the provided raw Ignition config
+func KubeletConfigFromRawIgnConfig(role, name string, rawIgnCfg []byte) (*mcfgv1.KubeletConfig, error) {
+	labels := map[string]string{
+		mcfgv1.KubeletConfigRoleLabelPrefix + role: "",
```
Are you labelling the kubeconfig pools.operator.machineconfiguration.openshift.io/master|worker? What happens for a custom pool?

kikisdeliveryservice replied:
Wouldn't a worker kubelet config would update a custom pool since custom pools inherit from workers.

sairameshv replied:
Right now, The workerlatencyprofiles are applied only on to the worker pool. Addressed the same in the code.

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,336 @@
+
+	for _, pool := range mcpPools {
+		role := pool.Name
+		// the configuration change will be applied only on the worker nodes.
```
Same question as above. Does the nodes.config.openshift.io resources only apply to workers? Or any custom pools?

kikisdeliveryservice replied:
custom pools inherit from worker pool so once a new worker mc is generated, they'd get this update too no? (unless I'm completely missing something here)

kikisdeliveryservice replied:
Ofc that said, then we should only be checking for the worker pool and not create the impression that one should be creating a seperate config for a custom pool.

yuqi-zhang commented on pkg/controller/kubelet-config/helpers.go:
```diff
@@ -227,6 +262,10 @@ func getManagedFeaturesKeyDeprecated(pool *mcfgv1.MachineConfigPool) string {
 }
 
+func getManagedNodeKey(pool *mcfgv1.MachineConfigPool, client mcfgclientset.Interface) (string, error) {
+	return ctrlcommon.GetManagedKey(pool, client, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
```
Is this using 99-kubelet-generated-xxx? Wouldn't this conflict with user supplied kubeconfigs?

sairameshv replied:
Yeah, updated the code to leverage MachineConfig resource instead of a kubeletConfig resource.

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,336 @@
+		}
+		if isNotFound {
+			ignConfig := ctrlcommon.NewIgnConfig()
+			kc, err = ctrlcommon.KubeletConfigFromIgnConfig(role, key, ignConfig)
```
I'm a bit confused as to this general flow. Am I understanding correctly that you create a kubeletconfig for the nodes resource under a specific key?

Since we are the kubeletconfigcontroller, shouldn't we be creating a new managed MachineConfig (generated via kubeletconfig if necessary in the controller) for nodes resources?

yuqi-zhang replied:
Another thing to consider is that each pool can only have 1 kubeletconfig. Meaning that you will either:
1. merge into user created ones (don't do this)
2. overwrite user created ones or have user created ones overwrite you, neither of which is an option

sairameshv replied:
> I'm a bit confused as to this general flow. Am I understanding correctly that you create a kubeletconfig for the nodes resource under a specific key?
> 
> Since we are the kubeletconfigcontroller, shouldn't we be creating a new managed MachineConfig (generated via kubeletconfig if necessary in the controller) for nodes resources?

@yuqi-zhang , I understand the concern and totally agree with your suggestion. My initial set of code changes were to create/update an MC whenever there is a change in the node config object. The only reason that I modified the code to create/update KC is to actually rely on it for the status updates.

I would update the code to create/update MC and would now rely on the corresponding pool status for updating the node config object's status. Should that be okay?

yuqi-zhang replied:
Hi Sai,

So to confirm, does the changes from the `node.config.openshift.io` modify the base kubelet files? If so, then I think the logic would have to be such that:

1. if there is the base system templates, create a 98-xxx MachineConfig (like we have today for featuregates) that merges in the base configs, maybe around https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/kubelet-config/kubelet_config_controller.go#L594 ?
2. if there is a featuregate, merge the featuregate changes and the nodes changes together with the base into 98-xx
3. if there is a user kubeletconfig, merge all of the above into 99-xxx
4. if there is a new user kubeletconfig, merge the new one, the featuregates, the node changes, and the base templates into 99-xxx-2

Basically, we need to make sure that features/featuregates/nodes config/user kubeletconfigs all get rendered together based on what combination of those exist. There must only be 1 MachineConfig for 

Conversely, if changes to node.config.openshift.io do NOT have to be merged into the base configs, this is much easier. You can e.g. create a MC that you manage that creates dropin files with unique names instead. This won't conflict.

Happy to discuss this synchronously if that would be helpful. Feel free to reach out to me on slack anytime, thanks!

yuqi-zhang reviewed (commented):
I think maybe I am confusing how the node resource is being applied. Looks like you create a kubeletconfig that the controller manages?

I think the controller should instead, like a featuregate, be create a machineconfig directly managed by the kubeletconfigcontroller. This additional abstraction will cause issues, e.g. with labelling and pools, and conflict with user provided kubeletconfigs heavily.

yuqi-zhang commented:
To clarify my previous concern, consider a system today just post-install. It has two MCs created by the KubeletConfigController:

01-master-kubelet 
01-worker-kubelet

That has the base configurations for the kubelet.

When you create a kubeletconfig as a user, we take that KC, merge it with the base 01-xx config, and create a new 99-generated-pool-xx machineconfig that essentially has both.

When you create a second kubeletconfig, we take the second KC and merge it with the base 01-xx config, and create a new 99-generated-pool-xx-2 config (note the suffix). This will not have the first user provided KC in it since we do no merging.

Thus the kubeletconfigcontroller operates under the model that it can never create KCs, otherwise it will disrupt the above flow.

So then the main KC-> MC constuction loop (that would create 01-xx, 99-generated-xx, 99-generated-xx-2/3/4/5...) needs to have an addendum that takes the base KC + whatever the node resource modifies for the kubelet + user provided kubeletconfig.

Now an interesting problem is, what happens if there is no user created KC. Should the node resource render go into 01-kubelet-xx?

1. if no, we risk conflicting with ourselves. We have other renders, e.g. 98-kubelet that is used for featureGates. Maybe this should also use 98-xx as a new MC if no KC exists?
2. if yes, we potentially risk messing with base templates, which other objects in this flow (featuregates, ICSP, etc.) doesn't do

I could just be overthinking the last bit, but the general idea applies

kikisdeliveryservice commented:
Discussed with Jerry and am in agreement that this work is getting done in the incorrect place in the flow and this PR needs to ensure that the nodeconfig changes get into the MC not the KC due to how KCs differ from MCs (no merging/user facing).
 
> Now an interesting problem is, what happens if there is no user created KC. Should the node resource render go into 01-kubelet-xx?
> 
>     1. if no, we risk conflicting with ourselves. We have other renders, e.g. 98-kubelet that is used for featureGates. Thinking about it, this can probably also render into the 98 (featuregate?) if both/either featuregate and node exists, but 99 (user) does not
> 
>     2. if yes, we potentially risk messing with base templates, which other objects in this flow (featuregates, ICSP, etc.) doesn't do
> 

Following the 98/99 pattern is imo more consistent. 01 isn't really a config that changes day2 (unless via upgrade) whereas we already have the 98/99 pattern for similar.

sairameshv commented:
Hello @kikisdeliveryservice , I've updated the code to leverage MC instead of KC. Discussed with @yuqi-zhang regarding his review comments and addressed all of them.
Could you PTAL now?

sairameshv commented:
/test bootstrap-unit

sairameshv commented:
/test bootstrap-unit

sairameshv commented:
/cc @kikisdeliveryservice @yuqi-zhang

rphillips commented on pkg/controller/kubelet-config/kubelet_config_nodes_test.go:
```diff
@@ -0,0 +1,114 @@
+		t.Run(string(platform), func(t *testing.T) {
+
+			cc := newControllerConfig(ctrlcommon.ControllerConfigName, platform)
+			mcp := helpers.NewMachineConfigPool("worker", nil, helpers.WorkerSelector, "v0")
```
We should add a test case for a worker and master.

sairameshv commented:
/test e2e-gcp-op

sairameshv commented:
/retest-required

sairameshv commented:
/retest-required

rphillips commented:
can the hold be taken off this PR?

yuqi-zhang commented on pkg/controller/kubelet-config/helpers.go:
```diff
@@ -227,6 +262,15 @@ func getManagedFeaturesKeyDeprecated(pool *mcfgv1.MachineConfigPool) string {
+}
+
+// Deprecated: use getManagedNodeKey
+func getManagedNodeConfigKeyDeprecated(pool *mcfgv1.MachineConfigPool) string {
```
I don't think you need to have this split out given there is no legacy case for node configs, but I think its not a big issue

sairameshv replied:
Addressed.

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,409 @@
+		}
+		nodeConfig, ok = tombstone.Obj.(*osev1.Node)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a KubeletConfig %#v", obj))
```
`KubeletConfig` -> `NodeConfig`

sairameshv replied:
Addressed

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,409 @@
+	} else {
+		nodeConfig.Status.WorkerLatencyProfileStatus.Conditions = append(nodeConfig.Status.WorkerLatencyProfileStatus.Conditions, nodeConfigCondition)
+	}
+	_, err := ctrl.configClient.ConfigV1().Nodes().UpdateStatus(context.TODO(), nodeConfig, metav1.UpdateOptions{})
```
Does anything else update the status? Could this be racy?

sairameshv replied:
This is the only place in the MCO where the NodeConfig status is getting updated

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,409 @@
+	return true
+}
+
+//
```
did you mean to comment something here

sairameshv replied:
Yes, Addressed

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,409 @@
+	ctrl.enqueueNodeConfig(nodeConfig)
+}
+
+func (ctrl *Controller) deleteNodeConfig(obj interface{}) {
```
Have you tested deleting a node config object? Does each MC correctly update? i.e. does the change propagate to the 98-xxx and 99-xxx ones?

sairameshv replied:
Yeah, this has been tested and verified. The change is getting reflected in case of the creation of kubeletconfig, featuregates.

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_controller.go:
```diff
@@ -507,6 +531,16 @@ func (ctrl *Controller) syncKubeletConfig(key string) error {
+	if macherrors.IsNotFound(err) {
+		nodeConfig = createNewDefaultNodeconfig()
+	} else if err != nil {
+		glog.V(2).Infof("%v", err)
```
V(2) is the default verbosity so this is always logged. Do you need to do that given that you are returning the error right below?

sairameshv replied:
Addressed

yuqi-zhang commented on pkg/controller/kubelet-config/helpers.go:
```diff
@@ -131,6 +132,40 @@ func createNewDefaultFeatureGate() *osev1.FeatureGate {
+}
+
+// updateOriginalKubeConfigwithNodeConfig updates the original Kubelet Configuration based on the Nodespecific configuration
+func updateOriginalKubeConfigwithNodeConfig(node *osev1.Node, originalKubeletConfig *kubeletconfigv1beta1.KubeletConfiguration) error {
```
So the interesting part is, since you are running this for every sync loop for KubeletConfig and features, what I think will happen is:

1. if you don't provide anything (no KC, no feature, no node), this code is not called at all
2. if you provide a kubeletconfig or feature on the system, this code is called, and the generated non-default will contain `originalKubeletConfig.NodeStatusUpdateFrequency = metav1.Duration{Duration: osev1.DefaultNodeStatusUpdateFrequency}`

so I guess the question then is, is there a difference between having `originalKubeletConfig.NodeStatusUpdateFrequency = metav1.Duration{Duration: osev1.DefaultNodeStatusUpdateFrequency}` and not having it? Because that is a behaviour change depending on whether you have other configs or not. If we want this to always be the default, we may need this somewhere else to generate it into the base.

Alternatively, if you covered this somewhere else and I am misunderstanding, please let me know where that happens. If this doesn't otherwise affect the default, you can ignore this question

yuqi-zhang reviewed (commented):
Did a pass and left some comments. The general direction is looking much better.

Also as a general note, would it be possible to structure your commits into 2:
1 - the API/client-go bump
2 - your code for the node object management

As a side note, this may also be a good candidate for the bootstrap-unit test, which was designed to test bootstrapping kubelet/features. But that doesn't have to be part of this PR

harche commented:
/test e2e-agnostic-upgrade

sairameshv commented:
/unhold

sairameshv commented on pkg/apis/machineconfiguration.openshift.io/v1/types.go:
```diff
@@ -321,6 +324,13 @@ type MachineConfigPoolCondition struct {
 	Message string `json:"message"`
 }
 
+const (
```
I thought this is the correct place as these constants can be referenced by any controller when there is a need to refer to the default machine config pool names.

yuqi-zhang replied:
Right, but this is the external exposed API interface. I would recommend, to share across controllers, to do so in `controller/common/constants.go`

yuqi-zhang replied:
We also have e.g. `masterPoolName = "master"` in node_controller.go's consts, which we can also delete in favour of common/constants

sairameshv replied:
Addressed

sairameshv commented on pkg/controller/common/helpers.go:
```diff
@@ -668,6 +668,35 @@ func MachineConfigFromRawIgnConfig(role, name string, rawIgnCfg []byte) (*mcfgv1
 }
 
+// KubeletConfigFromIgnConfig creates a KubeletConfig with the provided Ignition config
+func KubeletConfigFromIgnConfig(role, name string, ignCfg interface{}) (*mcfgv1.KubeletConfig, error) {
```
Yeah, removed the code.

sairameshv commented on pkg/controller/common/helpers.go:
```diff
@@ -668,6 +668,35 @@ func MachineConfigFromRawIgnConfig(role, name string, rawIgnCfg []byte) (*mcfgv1
+}
+
+// KubeletConfigFromRawIgnConfig creates a KubeletConfig with the provided raw Ignition config
+func KubeletConfigFromRawIgnConfig(role, name string, rawIgnCfg []byte) (*mcfgv1.KubeletConfig, error) {
```
Addressed

cgwalters commented:
I'll be honest I find this whole thing very confusing.  So we have a new nodes CRD which is generating kubeletconfigs which in turn generates machineconfig?

OK I see, it got changed to just directly generate machineconfig.  That helps.

Thinking about this in the context of [ocp coreos layering](https://github.com/openshift/enhancements/pull/1032)...we haven't yet fleshed out what it would look like to use machineconfig or kubeletconfig (or this new nodes CRD) as part of a `Dockerfile`.

But just imagine e.g.:
```
FROM quay.io/rhel-coreos:4.11
ADD kubelet-latency.conf /etc/kubelet.d
```

Where `kubelet-latency.conf` is a simple file that injects `--node-status-update-frequency 120` or whatever into the kubelet command line.

The towers of abstractions we've built up here would effectively disappear.   Now, yes there is the cost that one can inject arbitrary command line options and such and we do want to figure out a way to accept validated CRDs as part of a build processs or so but...still.

Anyways.  I didn't really review the original nodes CRD enhancement and I probably should have.

I know this is not the most useful feedback but I feel like we just need to think hard about how to simplify all this.  There are already so many sub-controllers and CRDs and flows inside the MCO and I find it hugely challenging to keep in my head, and this adds a whole new level to that.

sairameshv commented:
> Did a pass and left some comments. The general direction is looking much better.
> 
> Also as a general note, would it be possible to structure your commits into 2: 1 - the API/client-go bump 2 - your code for the node object management
> 
> As a side note, this may also be a good candidate for the bootstrap-unit test, which was designed to test bootstrapping kubelet/features. But that doesn't have to be part of this PR

Addressed, re-arranged the commits.

yuqi-zhang reviewed (commented):
So I did some testing with this PR:
1. create a nodes.config.openshift.io object with latency profile medium - result: created 97-generated-kubelet with `"nodeStatusUpdateFrequency": "20s",` :heavy_check_mark: 
2. create a kubeletconfig, generating 99-generated-kubelet with both the config and `"nodeStatusUpdateFrequency": "20s",` :heavy_check_mark: 
3. create another kubeletconfig, generating 99-generated-kubelet-1 with the new config and `"nodeStatusUpdateFrequency": "20s",` :heavy_check_mark: 
4. attempt to delete nodes.config.openshift.io/cluster - can't, which I think is expected :heavy_check_mark: 
5. attempt to change nodes.config.openshift.io/cluster back to latency profile default - the 97-generated-kubelet does work and changes to `"nodeStatusUpdateFrequency": "10s",` BUT the 99-kubelet-generated/99-kubelet-generated-1 still has `"nodeStatusUpdateFrequency": "20s",` . Meaning that it will overwrite (and thus will not work). i.e. if you have any kubeletconfigs on the node, you can't modify the profile anymore, which I think is incorrect :x: 
6. attempt to create a new nodes.config.openshift.io/test object. Creation is successful, but nothing happens. I assume only nodes.config.openshift.io/cluster is used? :grey_question: 

Re point 5: I think we have to somehow trigger the kubeletconfig object sync again with the latest kubeletconfig object such that it regenerates the config... which is problematic. Actually now I think about it, featuregates may also have this problem...? But that doesn't come up as often since you create them and don't generally change them, whereas the latency profile is multiple-step from default->medium->low.

yuqi-zhang commented:
One option for the above could be to have the node.config.openshift.io sync loop also check and update each kubeletconfig-machineconfig (i.e. 99-generated-worker-kubelet, 99-generated-worker-kubelet-1, etc.) and update all of them.

sairameshv commented:
> So I did some testing with this PR:
> 
> 1. create a nodes.config.openshift.io object with latency profile medium - result: created 97-generated-kubelet with `"nodeStatusUpdateFrequency": "20s",` heavy_check_mark
> 2. create a kubeletconfig, generating 99-generated-kubelet with both the config and `"nodeStatusUpdateFrequency": "20s",` heavy_check_mark
> 3. create another kubeletconfig, generating 99-generated-kubelet-1 with the new config and `"nodeStatusUpdateFrequency": "20s",` heavy_check_mark
> 4. attempt to delete nodes.config.openshift.io/cluster - can't, which I think is expected heavy_check_mark
> 5. attempt to change nodes.config.openshift.io/cluster back to latency profile default - the 97-generated-kubelet does work and changes to `"nodeStatusUpdateFrequency": "10s",` BUT the 99-kubelet-generated/99-kubelet-generated-1 still has `"nodeStatusUpdateFrequency": "20s",` . Meaning that it will overwrite (and thus will not work). i.e. if you have any kubeletconfigs on the node, you can't modify the profile anymore, which I think is incorrect x
> 6. attempt to create a new nodes.config.openshift.io/test object. Creation is successful, but nothing happens. I assume only nodes.config.openshift.io/cluster is used? grey_question
> 
> Re point 5: I think we have to somehow trigger the kubeletconfig object sync again with the latest kubeletconfig object such that it regenerates the config... which is problematic. Actually now I think about it, featuregates may also have this problem...? But that doesn't come up as often since you create them and don't generally change them, whereas the latency profile is multiple-step from default->medium->low.

Hello @yuqi-zhang , Thanks for bringing out this scenario.
Point 5:
The issue has been resolved by listing all the existing `kubeletconfigs` and calling the `syncKubeletConfig` inside the `syncNodeConfig` function.
Yeah, this is the issue with the existing featuregate sync as well. (I can add the fix as part of a new PR if required).

Point 6:
Reinforced the code to reject the addition/update of a new nodes.config.openshift.io object with the name other than "cluster". Tested and verified this functionality.

sairameshv commented:
/retest-required

sairameshv commented:
/retest-required

yuqi-zhang commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,434 @@
+		// skipping the update in case of the Worker-Latency-Profile type transition from "Default" to "LowUpdateSlowReaction" and vice-versa
+		// (TODO) Ideally the user request has to be honoured, the transition need to be from Default -> Medium -> Low or vice-versa.
+		// Restricting the request for now until this process is automated in future.
+		if (oldNode.Spec.WorkerLatencyProfile == osev1.DefaultUpdateDefaultReaction && newNode.Spec.WorkerLatencyProfile == osev1.LowUpdateSlowReaction) || (oldNode.Spec.WorkerLatencyProfile == osev1.LowUpdateSlowReaction && newNode.Spec.WorkerLatencyProfile == osev1.DefaultUpdateDefaultReaction) {
```
Hmm ok so this is the main check for whether the update is allowable or not.

So, I thought through the possibilities, and I think this will work so long as we only have 3 latency profiles.

i.e. this check is only for updates. So if I had 4 profiles that can only go from a->b->c->d and I want to skip b/c for example, I think can just do:
a->c (degrade, nothing happens)
c->d -> this actually works, because even though the actual config is on a, we are looking only at our object (on c) and desired (on d). So we essentially would have done a->d right? And this would let it through

In a more perfect scenario I think the controller creating the MC (sync function) should do the final check between current and desired. This should just enqueue and have the sync handle this.

I'm not against merging as is, but just for future, if we have more things to handle, the sync loop should be the place, and not individual informer checks.

sairameshv replied:
The idea here is to automate the process of this transition in future(as mentioned in the (TODO) comment) without rejecting the user request. That automatic transition needs a thorough design, testing and hence implemented a check here this way for now.

yuqi-zhang reviewed (commented):
Holding until https://github.com/openshift/machine-config-operator/pull/2868 merges. For consistency purposes, left a few nits below, as well as some other comments

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,434 @@
+	}
+
+	if ctrl.nodeConfigQueue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing node configuration %v: %v", key, err)
```
Addressed

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,434 @@
+		return err
+	}
+	// Fetch the Node
+	nodeConfig, err := ctrl.nodeConfigLister.Get(ctrlcommon.ClusterNodeInstanceName)
```
Yeah

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,434 @@
+		}
+		glog.Infof("Applied Node configuration %v on MachineConfigPool %v", key, pool.Name)
+		// fetch the kubeletconfigs
+		kcs, err := ctrl.mckLister.List(labels.Everything())
```
Sure, Addressed

sairameshv commented on pkg/controller/kubelet-config/kubelet_config_nodes.go:
```diff
@@ -0,0 +1,434 @@
+		}
+		for _, kc := range kcs {
+			// updating the existing kubeletconfigs with the updated nodeconfig
+			err := ctrl.syncKubeletConfig(kc.Name)
```
Yes

harche commented:
/retest

harche commented:
/test e2e-gcp-op

yuqi-zhang reviewed (approved):
There are a few more locations where there seems to be extra non-needed debug statements, as well as `%v` instead of `%w` in the error formatting, but let's fix those up as a follow up PR, no need to block the actual functionality on that.

Thanks for all the fixups. I think we are good to go.
